leg_detector/src/leg_detector_lib.cpp

```diff
@@ -0,0 +1,387 @@
+      double closest_dist = leg_pair_separation_m_;
+
+      for (leg1 = begin; leg1 != end; ++leg1)
+      {
+        // If this leg has an id or low reliability, skip
+        if ((*leg1)->object_id != ""
+            || (*leg1)->getReliability() < leg_reliability_limit_)
+          continue;
+
+        for (leg2 = begin; leg2 != end; ++leg2)
+        {
+          if (((*leg2)->object_id != "")
+              || ((*leg2)->getReliability() < leg_reliability_limit_)
+              || (leg1 == leg2)) continue;
+          double d = distance(leg1, leg2);
+          if (d < closest_dist)
+          {
+            best1 = leg1;
+            best2 = leg2;
+          }
+        }
+      }
+
+      if (best1 != end)
+      {
+        char id[100];
+        snprintf(id, 100, "Person%d", next_p_id_++);
+        (*best1)->object_id = std::string(id);
+        (*best2)->object_id = std::string(id);
+        (*best1)->other = *best2;
+        (*best2)->other = *best1;
+      }
+      else
+      {
+        break;
+      }
+    }
+  }
+
+  vector<people_msgs::PersonPositionMeasurement> LegDetector::processLaserScan(const sensor_msgs::LaserScan& scan, tf::TransformListener& tfl)
+  {
+    ScanProcessor processor(scan, mask_);
+
+    processor.splitConnected(connected_thresh_);
+    processor.removeLessThan(5);
+
+    cv::Mat tmp_mat = cv::Mat(1, feat_count_, CV_32FC1);
+
+    // if no measurement matches to a tracker in the last <no_observation_timeout>  seconds: erase tracker
+    ros::Time purge = scan.header.stamp + ros::Duration().fromSec(-no_observation_timeout_s_);
+    list<SavedFeature*>::iterator sf_iter = saved_features_.begin();
+    while (sf_iter != saved_features_.end())
+    {
+      if ((*sf_iter)->meas_time_ < purge)
+      {
+        if ((*sf_iter)->other)
+          (*sf_iter)->other->other = NULL;
+        delete(*sf_iter);
+        saved_features_.erase(sf_iter++);
+      }
+      else
+        ++sf_iter;
+    }
+
+
+    // System update of trackers, and copy updated ones in propagate list
+    list<SavedFeature*> propagated;
+    for (list<SavedFeature*>::iterator sf_iter = saved_features_.begin();
+         sf_iter != saved_features_.end();
+         sf_iter++)
+    {
+      (*sf_iter)->propagate(scan.header.stamp);
+      propagated.push_back(*sf_iter);
+    }
+
+
+    // Detection step: build up the set of "candidate" clusters
+    // For each candidate, find the closest tracker (within threshold) and add to the match list
+    // If no tracker is found, start a new one
+    multiset<MatchedFeature> matches;
+    for (list<SampleSet*>::iterator i = processor.getClusters().begin();
+         i != processor.getClusters().end();
+         i++)
+    {
+      vector<float> f = calcLegFeatures(*i, scan);
+
+      for (int k = 0; k < feat_count_; k++)
+        tmp_mat.data[k] = (float)(f[k]);
+
+      float probability = 0.5 - forest->predict(tmp_mat, cv::noArray(), cv::ml::RTrees::PREDICT_SUM) / forest->getRoots().size();
```

Comment:
getRoots cannot be zero because if it is, in that learned model is messed up which makes this anyways unusable.